README.md
# Vagrant Google Compute Engine (GCE) Provider

[![Gem Version](https://badge.fury.io/rb/vagrant-google.svg)](https://badge.fury.io/rb/vagrant-google)

[gem]: https://rubygems.org/gems/vagrant-google
[gemnasium]: https://gemnasium.com/mitchellh/vagrant-google

This is a [Vagrant](http://www.vagrantup.com) 2.0.3+ plugin that adds an
[Google Compute Engine](http://cloud.google.com/compute/) (GCE) provider to
Vagrant, allowing Vagrant to control and provision instances in GCE.

The maintainers for this plugin are @temikus(primary), @erjohnso(backup).

## Features

* Boot Google Compute Engine instances.
* SSH into the instances.
* Provision the instances with any built-in Vagrant provisioner.
* Synced folder support via Vagrant's
[rsync action](https://www.vagrantup.com/docs/synced-folders/rsync.html).
* Define zone-specific configurations so Vagrant can manage machines in
  multiple zones.

## Usage

Install using standard Vagrant 1.1+ plugin installation methods.  After
installing, `vagrant up` and specify the `google` provider.  For example,

```sh
$ vagrant plugin install vagrant-google
...
$ vagrant up --provider=google
...
```

Of course, prior to this you'll need to obtain a GCE-compatible box file for
Vagrant. You may also need to ensure you have a ruby-dev and other utilities
such as GNU make installed prior to installing the plugin.

## Google Cloud Platform Setup

Prior to using this plugin, you will first need to make sure you have a
Google Cloud Platform account, enable Google Compute Engine, and create a
Service Account for API Access.

1. Log in with your Google Account and go to
   [Google Cloud Platform](https://cloud.google.com) and click on the
   `Try it free` button.
1. Create a new project and remember to record the `Project ID`
1. Next, enable the
   [Google Compute Engine API](https://console.cloud.google.com/apis/api/compute_component/)
   for your project in the API console. If prompted, review and agree to the
   terms of service.
1. While still in the API Console, go to
   [Credentials subsection](https://console.cloud.google.com/apis/credentials),
   and click `Create credentials` -> `Service account key`. In the
   next dialog, create a new service account, select `JSON` key type and
   click `Create`.
1. Download the JSON private key and save this file in a secure
   and reliable location.  This key file will be used to authorize all API
   requests to Google Compute Engine.
1. Still on the same page, click on
   [Manage service accounts](https://console.cloud.google.com/permissions/serviceaccounts)
   link to go to IAM console. Copy the `Service account id` value of the service
   account you just selected. (it should end with `gserviceaccount.com`) You will
   need this email address and the location of the private key file to properly
   configure this Vagrant plugin.
1. Add the SSH key you're going to use to GCE Metadata in `Compute` ->
   `Compute Engine` -> `Metadata` section of the console, `SSH Keys` tab. (Read
   the [SSH Support](https://github.com/mitchellh/vagrant-google#ssh-support)
   readme section for more information.)

## Quick Start

After installing the plugin (instructions above), the quickest way to get
started is to actually use a dummy Google box from Atlas and specify all the
details manually within a `config.vm.provider` block.

So first, make a Vagrantfile that looks like the following, filling in
your information where necessary:

```ruby
Vagrant.configure("2") do |config|
  config.vm.box = "google/gce"

  config.vm.provider :google do |google, override|
    google.google_project_id = "YOUR_GOOGLE_CLOUD_PROJECT_ID"
    google.google_json_key_location = "/path/to/your/private-key.json"

    google.image_family = 'ubuntu-1604-lts'

    override.ssh.username = "USERNAME"
    override.ssh.private_key_path = "~/.ssh/id_rsa"
    #override.ssh.private_key_path = "~/.ssh/google_compute_engine"
  end

end
```

And then run `vagrant up --provider=google`.

This will start a latest version of Ubuntu 16.04 LTS instance in the
`us-central1-f` zone, with an `n1-standard-1` machine, and the `"default"`
network within your project. And assuming your SSH information (see below) was
filled in properly within your Vagrantfile, SSH and provisioning will work as
well.

Note that normally a lot of this boilerplate is encoded within the box file,
but the box file used for the quick start, the "google" box, has no
preconfigured defaults.

## SSH Support

In order for SSH to work properly to the GCE VM, you will first need to add
your public key to the GCE metadata service for the desired VM user account.
When a VM first boots, a Google-provided daemon is responsible for talking to
the internal GCE metadata service and creates local user accounts and their
respective `~/.ssh/authorized_keys` entries.  Most new GCE users will use the
[Cloud SDK](https://cloud.google.com/sdk/) `gcloud compute` utility when first
getting started with GCE. This utility has built in support for creating SSH
key pairs, and uploading the public key to the GCE metadata service.  By
default, `gcloud compute` creates a key pair named
`~/.ssh/google_compute_engine[.pub]`.

Note that you can use the more standard `~/.ssh/id_rsa[.pub]` files, but you
will need to manually add your public key to the GCE metadata service so your
VMs will pick up the the key. Note that they public key is typically
prefixed with the username, so that the daemon on the VM adds the public key
to the correct user account.

Additionally, you will probably need to add the key and username to override
settings in your Vagrantfile like so:

```ruby
config.vm.provider :google do |google, override|

    #...google provider settings are skipped...

    override.ssh.username = "testuser"
    override.ssh.private_key_path = "~/.ssh/id_rsa"

    #...google provider settings are skipped...

end
```

See the links below for more help with SSH and GCE VMs.

  * https://cloud.google.com/compute/docs/instances#sshing
  * https://cloud.google.com/compute/docs/console#sshkeys

## Box Format

Every provider in Vagrant must introduce a custom box format. This provider
introduces `google` boxes. You can view an example box in
[example_boxes/](https://github.com/mitchellh/vagrant-google/tree/master/example_boxes).
That directory also contains instructions on how to build a box.

The box format is basically just the required `metadata.json` file along with
a `Vagrantfile` that does default settings for the provider-specific
configuration for this provider.

## Configuration

This provider exposes quite a few provider-specific configuration options:

* `google_json_key_location` - The location of the JSON private key file matching your
  Service Account.
  (Can also be configured with `GOOGLE_JSON_KEY_LOCATION` environment variable.)
* `google_project_id` - The Project ID for your Google Cloud Platform account.
  (Can also be configured with `GOOGLE_PROJECT_ID` environment variable.)
* `image` - The image name to use when booting your instance.
* `image_family` - Specify an "image family" to pull the latest image from. For example: `centos-7`
will pull the most recent CentOS 7 image. For more info, refer to
[Google Image documentation](https://cloud.google.com/compute/docs/images#image_families).
* `image_project_id` - The ID of the GCP project to search for the `image` or `image_family`.
* `instance_group` - Unmanaged instance group to add the machine to. If one
  doesn't exist it will be created.
* `instance_ready_timeout` - The number of seconds to wait for the instance
  to become "ready" in GCE. Defaults to 20 seconds.
* `machine_type` - The machine type to use.  The default is "n1-standard-1".
* `disk_size` - The disk size in GB.  The default is 10.
* `disk_name` - The disk name to use.  If the disk exists, it will be reused, otherwise created.
* `disk_type` - Whether to use Standard disk or SSD disk. Use either `pd-ssd` or `pd-standard`.
* `autodelete_disk` - Boolean whether to delete the disk when the instance is deleted or not. Default is true.
* `metadata` - Custom key/value pairs of metadata to add to the instance.
* `name` - The name of your instance.  The default is "i-yyyymmddhh-randomsd",
  e.g. 10/08/2015 13:15:15 is "i-2015081013-15637fda".
* `network` - The name of the network to use for the instance.  Default is
 "default".
* `network_project_id` - The ID of the GCP project for the network and subnetwork to use for the instance. Default is `google_project_id`.
* `subnetwork` - The name of the subnetwork to use for the instance.
* `tags` - An array of tags to apply to this instance.
* `labels` - Custom key/value pairs of labels to add to the instance.
* `zone` - The zone name where the instance will be created.
* `can_ip_forward` - Boolean whether to enable IP Forwarding.
* `external_ip` - The external IP address to use (supports names). Set to `false` to not assign an external address.
* `network_ip` - The internal IP address to use. Default is to use next available address.
* `use_private_ip` - Boolean whether to use private IP for SSH/provisioning. Default is false.
* `preemptible` - Boolean whether to enable preemptibility. Default is false.
* `auto_restart` - Boolean whether to enable auto_restart. Default is true.
* `on_host_maintenance` - What to do on host maintenance. Can be set to `MIGRATE` or `TERMINATE` Default is `MIGRATE`.
* `scopes` or `service_accounts` - An array of OAuth2 account scopes for
  services that the instance will have access to. Those can be both full API
  scopes, just endpoint aliases (the part after `...auth/`), and `gcloud`
  utility aliases, for example:
  `['storage-full', 'bigquery', 'https://www.googleapis.com/auth/compute']`.
* `service_account` - The IAM service account email to use for the instance.
* `additional_disks` - An array of additional disk configurations. `disk_size` is default to `10`GB;
  `disk_name` is default to `name` + "-additional-disk-#{index}"; `disk_type` is default to `pd-standard`;
  `autodelete_disk` is default to `true`. Here is an example of configuration.
  ```ruby
    [{
     :image_family => "google-image-family",
     :image => nil,
     :image_project_id => "google-project-id",
     :disk_size => 20,
     :disk_name => "google-additional-disk-0",
     :disk_type => "pd-standard",
     :autodelete_disk => true
    }]
  ```
* `accelerators` - An array of accelerator configurations. `type` is the
  accelerator type (e.g. `nvidia-tesla-k80`); `count` is the number of
  accelerators and defaults to 1. Note that only `TERMINATE` is supported for
  `on_host_maintenance`; this should be set explicitly, since the default is
  `MIGRATE`.
  ```ruby
  google.accelerators = [{
    :type => "nvidia-tesla-k80",
    :count => 2
  }]

  google.on_host_maintenance = "TERMINATE"
  ```
* `enable_secure_boot` - For [Shielded VM](https://cloud.google.com/security/shielded-cloud/shielded-vm), whether to enable Secure Boot.
* `enable_vtpm` - For [Shielded VM](https://cloud.google.com/security/shielded-cloud/shielded-vm), whether to enable vTPM.
* `enable_integrity_monitoring` - For [Shielded VM](https://cloud.google.com/security/shielded-cloud/shielded-vm), whether to enable Integrity monitoring.

These can be set like typical provider-specific configuration:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :google do |google|
    google.google_project_id = "YOUR_GOOGLE_CLOUD_PROJECT_ID"
    google.google_json_key_location = "/path/to/your/private-key.json"
  end
end
```

In addition to the above top-level configs, you can use the `zone_config`
method to specify zone-specific overrides within your Vagrantfile. Note
that the top-level `zone` config must always be specified to choose which
zone you want to actually use, however. This looks like this:

```ruby
Vagrant.configure("2") do |config|

  config.vm.box = "google/gce"

  config.vm.provider :google do |google|
    google.google_project_id = "YOUR_GOOGLE_CLOUD_PROJECT_ID"
    google.google_json_key_location = "/path/to/your/private-key.json"

    # Make sure to set this to trigger the zone_config
    google.zone = "us-central1-f"

    google.zone_config "us-central1-f" do |zone1f|
        zone1f.name = "testing-vagrant"
        zone1f.image = "debian-9-stretch-v20180611"
        zone1f.machine_type = "n1-standard-4"
        zone1f.zone = "us-central1-f"
        zone1f.metadata = {'custom' => 'metadata', 'testing' => 'foobarbaz'}
        zone1f.scopes = ['bigquery', 'monitoring', 'https://www.googleapis.com/auth/compute']
        zone1f.tags = ['web', 'app1']
    end
  end
end
```

The zone-specific configurations will override the top-level configurations
when that zone is used. They otherwise inherit the top-level configurations,
as you would expect.

There are a few example Vagrantfiles located in the
[vagrantfile_examples/ directory](https://github.com/mitchellh/vagrant-google/tree/master/vagrantfile_examples/).

## Networks

Networking features in the form of `config.vm.network` are not supported

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is loosely based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Next


## 2.7.0 (November 2021)

### User-facing

#### Added

- \#255 Adds enable_display option [dcode]
- \#248 Use google_application_default credentials by default [pecigonzalo]

## 2.6.0 (December 2020)

### User-facing

#### Added

- \#248 Add support for modifying the configuration of Shielded VM [lcy0321]
- \#244 Add accelerator (GPU) configuration [johanvdhaegen]
- \#216 Add support for Windows hosts through WinRM [dvanbrug]

#### Fixed

- \#246 Fix config override during metadata merge [mavin]
- \#249 Fix shielded API failures on non-shielded VM's [temikus]
- \#227 Fix additional disks being auto-deleted unless a disk type is specified [chrisgavin]

### Development

#### Fixed

- \#247 chore: Update Gemfile with new `vagrant-spec` branch name [mavin]
- \#241 Fix unfortunate CI `on:` trigger typo [temikus]
- \#240 Update development environment, add CI functionality [Temikus]
- \#239 Remove old fixup in Gemfile [Temikus]
- \#237 Bump image size for test skeleton [Temikus]


## 2.5.0 (September 2019)

### User-facing

#### Added
- \#222 Added internal IP support [andrewh1978]

#### Deprecated

- \#220 Deprecated google_client_email parameter, as it's no longer needed or
  supported by fog-google [temikus]

### Development

#### Added

- \#223 Set up unit test GitHub Action worflow [temikus]

#### Fixed

- \#225 Locked the vagrant dep to 2.2.4 due to bundler issues on 2.2.5 [temikus]

## 2.4.0 (April 2019)

### User-facing

#### Added
- \#213 Implemented Application Default Credentials authentication [mavin]

#### Fixed
- \#214 Set a default zone only if `default` network is used [mavin]
- \#215 Allow tags,labels and additional_disks to be merged with multiple
  configs [mavin]

### Development

- \#213 Bumped dependencies [mavin]
   - fog-google version to 1.9.0

## 2.3.0 (February 2019)

### User-facing

- \#210 Allow adding additional disks to the instances. [whynick1]

### Development

- \#211 Rspec-its is now explicitly required for unit tests. [temikus]

## 2.2.1 (October 2018)

### User-facing

- \#206 Fix image selection logic - Plugin no longer traces back with
  `image_family` config option. [temikus]

### Development

- \#206 Bumped dependencies. [temikus]
  - fog-google version to 1.8.1
  - vagrant & vagrant-spec are now pointing to new upstream Hashicorp org repos

## 2.2.0 (June 2018)

#### Fixed
* Bumped fog-google to v1.4.
This is a necessary upstream update to work properly with Ruby 2.4+ on some
platforms.

## 2.1.0 (May 2018)

* Add new configuration option `image_project_id` to allow using GCE images from other projects. [seanmalloy]
* Add new configuration option `network_project_id` to allow using GCP Shared VPC networks. [seanmalloy]
* Add new configuration option `service_account` to allow setting the IAM service account on instances. [seanmalloy]
* Deprecate configuration option `service_accounts`. Use `scopes` configuration option instead. [seanmalloy]

## 2.0.0 (March 2018)

* Update to use fog-google gem v1
* Add new configuration option `labels` for setting [labels](https://cloud.google.com/compute/docs/labeling-resources)
  on GCE instances
* Fix disk cleanup issue causing the disk to be marked as created before insertion
* Test environment fixups to avoid 'Encoded files can't be read outside of the Vagrant installer.'
* Breaking changes:
  * Drop support for configuration option `google_key_location`(GCP P12 key)
  * `image` parameter no longer defaults to an arbitrary image and must be
    specified at runtime
  * Rsync behavior now consistent with Vagrant's default, removed old rsync code

## 1.0.0 (July 2017)
## 0.2.5 (October 2016)
## 0.2.4 (April 2016)
## 0.2.3 (January 2016)

## 0.2.2 (October 2015)

* Cleanup instance and disks on backend failures [p0deje]
* Refactoring ssh warnings into separate action [temikus]
* Refactoring disk type detection logic [temikus]
* Miscellaneous doc updates and minor fixes [mbrukman, temikus]

## 0.2.1 (July 2015)

* Temporarily reverted the old SyncedFolders behaviour. (See #94)

## 0.2.0 (July 2015)

* Added support for service account definitions [tcr]
* Added support for preemptible instances [jcdang]
* Implemented auto_restart and on_host_maintenance options [jcdang]
* Implemented vagrant halt and reload actions [temikus]
* Added support for IP address specification by name [temikus]
* Instance name now defaults to time + uuid [temikus]
* Removed legacy rsync code, switched to Vagrant built-in SyncedFolders [temikus]
* Switched to fog-google metagem [temikus]
* Added a linter and custom acceptance tests [temikus]
* Updated documentation and examples [mbrukman, temikus]
* Miscellaneous UI/UX updates and bugfixes [temikus]

## 0.1.5 (May 2015)

* Added support for JSON private keys [temikus]
* Added disk_type parameter support [temikus]
* Added acceptance tests [temikus]
* Added can_ip_forward, external_ip, autodelete_disk and disk_name parameters support [phueper]
* Added support for user specified rsync excludes [patkar]
* Miscellaneous bugfixes [mbrukman, beauzeaux, iceydee, mklbtz, temikus]

## 0.1.4 (October 2014)

* Add option for disk size [franzs]
* Add tags [ptone]
* Updated default for latest Debian image

## 0.1.3 (July 2014)

* Updated all image references
* Fixed fog deprecation warning
* Updated example box `google.box`
* Got spec tests passing again

## 0.1.1 (October 11, 2013)

* Fixed bug with instance ready/SSH

## 0.1.0 (August 14, 2013)

* Initial release.


lib/vagrant-google/config.rb
# Copyright 2013 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
require "vagrant"
require "securerandom"

module VagrantPlugins
  module Google
    class Config < Vagrant.plugin("2", :config) # rubocop:disable Metrics/ClassLength
      # The path to the Service Account json-formatted private key
      #
      # @return [String]
      attr_accessor :google_json_key_location

      # The Google Cloud Project ID (not name or number)
      #
      # @return [String]
      attr_accessor :google_project_id

      # The image name of the instance to use.
      #
      # @return [String]
      attr_accessor :image

      # The image family of the instance to use.
      #
      # @return [String]
      attr_accessor :image_family

      # The name of the image_project_id
      #
      # @return [String]
      attr_accessor :image_project_id

      # The instance group name to put the instance in.
      #
      # @return [String]
      attr_accessor :instance_group

      # The type of machine to launch, such as "n1-standard-1"
      #
      # @return [String]
      attr_accessor :machine_type

      # The size of disk in GB
      #
      # @return [Int]
      attr_accessor :disk_size

      # The name of the disk to be used, it it exists, it will be reused, otherwise created.
      #
      # @return [String]
      attr_accessor :disk_name

      # The type of the disk to be used, such as "pd-standard"
      #
      # @return [String]
      attr_accessor :disk_type

      # The user metadata string
      #
      # @return [Hash<String, String>]
      attr_accessor :metadata

      # The name of the instance
      #
      # @return [String]
      attr_accessor :name

      # The name of the network
      #
      # @return [String]
      attr_accessor :network

      # The name of the network_project_id
      #
      # @return [String]
      attr_accessor :network_project_id

      # The name of the subnetwork
      #
      # @return [String]
      attr_accessor :subnetwork

      # Tags to apply to the instance
      #
      # @return [Array]
      attr_accessor :tags

      # Labels to apply to the instance
      #
      # @return [Hash<String, String>]
      attr_accessor :labels

      # whether to enable ip forwarding
      #
      # @return Boolean
      attr_accessor :can_ip_forward

      # The external IP Address to use
      #
      # @return String
      attr_accessor :external_ip

      # The network IP Address to use
      #
      # @return String
      attr_accessor :network_ip

      # Use private ip address
      #
      # @return Boolean
      attr_accessor :use_private_ip

      # whether to autodelete disk on instance delete
      #
      # @return Boolean
      attr_accessor :autodelete_disk

      # Availability policy
      # whether to run instance as preemptible
      #
      # @return Boolean
      attr_accessor :preemptible

      # Availability policy
      # whether to have instance restart on failures
      #
      # @return Boolean
      attr_accessor :auto_restart

      # Availability policy
      # specify what to do when infrastructure maintenance events occur
      # Options: MIGRATE, TERMINATE
      # The default is MIGRATE.
      #
      # @return String
      attr_accessor :on_host_maintenance

      # The timeout value waiting for instance ready
      #
      # @return [Int]
      attr_accessor :instance_ready_timeout

      # The zone to launch the instance into.
      # If nil and the "default" network is set use the default us-central1-f.
      #
      # @return [String]
      attr_accessor :zone

      # The list of access scopes for instance.
      #
      # @return [Array]
      attr_accessor :scopes

      # Deprecated: the list of access scopes for instance.
      #
      # @return [Array]
      attr_accessor :service_accounts

      # IAM service account for instance.
      #
      # @return [String]
      attr_accessor :service_account

      # The configuration for additional disks.
      #
      # @return [Array<Hash>]
      attr_accessor :additional_disks

      # (Optional - Override default WinRM setup before for Public Windows images)
      #
      # @return [Boolean]
      attr_accessor :setup_winrm_password

      # Accelerators
      #
      # @return [Array<Hash>]
      attr_accessor :accelerators

      # whether the instance has Secure Boot enabled
      #
      # @return Boolean
      attr_accessor :enable_secure_boot

      # whether the instance has a display enabled
      #
      # @return Boolean
      attr_accessor :enable_display

      # whether the instance has the vTPM enabled
      #
      # @return Boolean
      attr_accessor :enable_vtpm

      # whether the instance has integrity monitoring enabled
      #
      # @return Boolean
      attr_accessor :enable_integrity_monitoring

      def initialize(zone_specific=false)
        @google_json_key_location    = UNSET_VALUE
        @google_project_id           = UNSET_VALUE
        @image                       = UNSET_VALUE
        @image_family                = UNSET_VALUE
        @image_project_id            = UNSET_VALUE
        @instance_group              = UNSET_VALUE
        @machine_type                = UNSET_VALUE
        @disk_size                   = UNSET_VALUE
        @disk_name                   = UNSET_VALUE
        @disk_type                   = UNSET_VALUE
        @metadata                    = {}
        @name                        = UNSET_VALUE
        @network                     = UNSET_VALUE
        @network_project_id          = UNSET_VALUE
        @subnetwork                  = UNSET_VALUE
        @tags                        = []
        @labels                      = {}
        @can_ip_forward              = UNSET_VALUE
        @external_ip                 = UNSET_VALUE
        @network_ip                  = UNSET_VALUE
        @use_private_ip              = UNSET_VALUE
        @autodelete_disk             = UNSET_VALUE
        @preemptible                 = UNSET_VALUE
        @auto_restart                = UNSET_VALUE
        @on_host_maintenance         = UNSET_VALUE
        @instance_ready_timeout      = UNSET_VALUE
        @zone                        = UNSET_VALUE
        @scopes                      = UNSET_VALUE
        @service_accounts            = UNSET_VALUE
        @service_account             = UNSET_VALUE
        @additional_disks            = []
        @setup_winrm_password        = UNSET_VALUE
        @accelerators                = []
        @enable_secure_boot          = UNSET_VALUE
        @enable_display              = UNSET_VALUE
        @enable_vtpm                 = UNSET_VALUE
        @enable_integrity_monitoring = UNSET_VALUE

        # Internal state (prefix with __ so they aren't automatically
        # merged)
        @__compiled_zone_configs = {}
        @__finalized = false
        @__zone_config = {}
        @__zone_specific = zone_specific
      end

      # Allows zone-specific overrides of any of the settings on this
      # configuration object. This allows the user to override things like
      # image and machine type name for zones. Example:
      #
      #     google.zone_config "us-central1-f" do |zone|
      #       zone.image = "ubuntu-1604-xenial-v20180306"
      #       zone.machine_type = "n1-standard-4"
      #     end
      #
      # @param [String] zone The zone name to configure.
      # @param [Hash] attributes Direct attributes to set on the configuration
      #   as a shortcut instead of specifying a full block.
      # @yield [config] Yields a new Google configuration.
      def zone_config(zone, attributes=nil, &block)
        # Append the block to the list of zone configs for that zone.
        # We'll evaluate these upon finalization.
        @__zone_config[zone] ||= []

        # Append a block that sets attributes if we got one
        if attributes
          attr_block = lambda do |config|
            config.set_options(attributes)
          end

          @__zone_config[zone] << attr_block
        end

        # Append a block if we got one
        @__zone_config[zone] << block if block_given?
      end

      #-------------------------------------------------------------------
      # Internal methods.
      #-------------------------------------------------------------------

      def merge(other)
        super.tap do |result|
          # Copy over the zone specific flag. "True" is retained if either
          # has it.
          new_zone_specific = other.instance_variable_get(:@__zone_specific)
          result.instance_variable_set(
            :@__zone_specific, new_zone_specific || @__zone_specific
          )

          # Go through all the zone configs and prepend ours onto
          # theirs.
          new_zone_config = other.instance_variable_get(:@__zone_config)
          @__zone_config.each do |key, value|
            new_zone_config[key] ||= []
            new_zone_config[key] = value + new_zone_config[key]
          end

          # Set it
          result.instance_variable_set(:@__zone_config, new_zone_config)

          # Merge in the metadata
          result.metadata = self.metadata.merge(other.metadata)

          # Merge in the labels
          result.labels = self.labels.merge(other.labels)

          # Merge in the tags
          result.tags |= self.tags
          result.tags |= other.tags

          # Merge in the additional disks
          result.additional_disks |= self.additional_disks
          result.additional_disks |= other.additional_disks
        end
      end

      def finalize! # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
        # Try to get access keys from standard Google environment variables; they
        # will default to nil if the environment variables are not present.
        @google_json_key_location = ENV['GOOGLE_JSON_KEY_LOCATION'] if @google_json_key_location == UNSET_VALUE
        @google_project_id = ENV['GOOGLE_PROJECT_ID'] if @google_project_id == UNSET_VALUE

        # Default image is nil
        @image = nil if @image == UNSET_VALUE

        # Default image family is nil
        @image_family = nil if @image_family == UNSET_VALUE

        # Default image project is nil
        @image_project_id = nil if @image_project_id == UNSET_VALUE

        # Default instance group name is nil
        @instance_group = nil if @instance_group == UNSET_VALUE

        # Default instance type is an n1-standard-1
        @machine_type = "n1-standard-1" if @machine_type == UNSET_VALUE

        # Default disk size is 10 GB
        @disk_size = 10 if @disk_size == UNSET_VALUE

        # Default disk name is nil
        @disk_name = nil if @disk_name == UNSET_VALUE

        # Default disk type is pd-standard
        @disk_type = "pd-standard" if @disk_type == UNSET_VALUE

        # Instance name defaults to a new datetime value + random seed
        # e.g. i-2015081013-15637fdb
        if @name == UNSET_VALUE
          t = Time.now
          @name = "i-#{t.strftime("%Y%m%d%H")}-" + SecureRandom.hex(4)
        end

        # Network defaults to 'default'
        @network = "default" if @network == UNSET_VALUE

        # Network project id defaults to project_id
        @network_project_id = @google_project_id if @network_project_id == UNSET_VALUE

        # Subnetwork defaults to nil
        @subnetwork = nil if @subnetwork == UNSET_VALUE

        # Default zone is us-central1-f if using the default network
        if @zone == UNSET_VALUE
          @zone = nil
          if @network == "default"
            @zone = "us-central1-f"
          end
        end

        # autodelete_disk defaults to true
        @autodelete_disk = true if @autodelete_disk == UNSET_VALUE

        # can_ip_forward defaults to nil
        @can_ip_forward = nil if @can_ip_forward == UNSET_VALUE

        # external_ip defaults to nil
        @external_ip = nil if @external_ip == UNSET_VALUE

        # network_ip defaults to nil
        @network_ip = nil if @network_ip == UNSET_VALUE

        # use_private_ip defaults to false
        @use_private_ip = false if @use_private_ip == UNSET_VALUE

        # preemptible defaults to false
        @preemptible = false if @preemptible == UNSET_VALUE

        # auto_restart defaults to true
        @auto_restart = true if @auto_restart == UNSET_VALUE

        # on_host_maintenance defaults to MIGRATE
        @on_host_maintenance = "MIGRATE" if @on_host_maintenance == UNSET_VALUE

        # Default instance_ready_timeout
        @instance_ready_timeout = 20 if @instance_ready_timeout == UNSET_VALUE

        # Default access scopes
        @scopes = nil if @scopes == UNSET_VALUE

        # Default access scopes
        @service_accounts = nil if @service_accounts == UNSET_VALUE

        # Default IAM service account
        @service_account = nil if @service_account == UNSET_VALUE

        # Default Setup WinRM Password
        @setup_winrm_password = nil if @setup_winrm_password == UNSET_VALUE

        # Config option service_accounts is deprecated
        if @service_accounts
          @scopes = @service_accounts
        end

        # enable_secure_boot defaults to nil
        @enable_secure_boot = false if @enable_secure_boot == UNSET_VALUE

        # enable_display defaults to nil
        @enable_display = false if @enable_display == UNSET_VALUE

        # enable_vtpm defaults to nil
        @enable_vtpm = false if @enable_vtpm == UNSET_VALUE

        # enable_integrity_monitoring defaults to nil
        @enable_integrity_monitoring = false if @enable_integrity_monitoring == UNSET_VALUE

        # Compile our zone specific configurations only within
        # NON-zone-SPECIFIC configurations.
        unless @__zone_specific
          @__zone_config.each do |zone, blocks|
            config = self.class.new(true).merge(self)

            # Execute the configuration for each block
            blocks.each { |b| b.call(config) }

            # The zone name of the configuration always equals the
            # zone config name:
            config.zone = zone

            # Finalize the configuration
            config.finalize!

            # Store it for retrieval
            @__compiled_zone_configs[zone] = config
          end
        end

        # Mark that we finalized
        @__finalized = true
      end

      def validate(machine)
        errors = _detected_errors

        errors << I18n.t("vagrant_google.config.zone_required") if @zone.nil?

        if @zone
          config = get_zone_config(@zone)

          # TODO: Check why provider-level settings are validated in the zone config
          errors << I18n.t("vagrant_google.config.google_project_id_required") if \
            config.google_project_id.nil?

          if config.google_json_key_location
            errors << I18n.t("vagrant_google.config.private_key_missing") unless \
              File.exist?(File.expand_path(config.google_json_key_location.to_s)) or
              File.exist?(File.expand_path(config.google_json_key_location.to_s, machine.env.root_path))
          end

          if config.preemptible
            errors << I18n.t("vagrant_google.config.auto_restart_invalid_on_preemptible") if \
             config.auto_restart
            errors << I18n.t("vagrant_google.config.on_host_maintenance_invalid_on_preemptible") unless \
             config.on_host_maintenance == "TERMINATE"
          end

          if config.image_family
            errors << I18n.t("vagrant_google.config.image_and_image_family_set") if \
             config.image
          end

          errors << I18n.t("vagrant_google.config.image_required") if config.image.nil? && config.image_family.nil?
          errors << I18n.t("vagrant_google.config.name_required") if @name.nil?

          if !config.accelerators.empty?
            errors << I18n.t("vagrant_google.config.on_host_maintenance_invalid_with_accelerators") unless \
              config.on_host_maintenance == "TERMINATE"
          end
        end

        if @service_accounts
          machine.env.ui.warn(I18n.t("vagrant_google.config.service_accounts_deprecaated"))
        end

        { "Google Provider" => errors }
      end

      # This gets the configuration for a specific zone. It shouldn't
      # be called by the general public and is only used internally.
      def get_zone_config(name)
        unless @__finalized
          raise "Configuration must be finalized before calling this method."
        end

        # Return the compiled zone config
        @__compiled_zone_configs[name] || self
      end
    end
  end
end
